src/main/ml-modules/services/KoopProvider.sjs

```diff
@@ -140,46 +140,90 @@ function generateServiceDescriptor(serviceName) {
 }
 
 function generateFieldDescriptors(layerModel, serviceName) {
+  if (layerModel.view === undefined) {
+    return generateFieldDescriptorsFromDataSourcesArray(layerModel, serviceName);
+  } else {
+    return generateFieldDescriptorsFromViewAndJoins(layerModel, serviceName);
+  }
+}
+
+function generateFieldDescriptorsFromViewAndJoins(layerModel, serviceName) {
   const fields = [];
 
-  const schema = getSchema(layerModel, serviceName);
-  const viewDef = tde.getView(schema, layerModel.view);
+  var schema = getSchema(layerModel, serviceName);
+  var view = layerModel.view;
+  const viewDef = tde.getView(schema, view);
+  generateFieldDescriptorsFromViewDef(viewDef, fields);
 
-  viewDef.view.columns.forEach((c) => {
-    const field = {
-      name : c.column.name,
-      type : getFieldType(c.column.scalarType)
-    };
+  if (layerModel.joins) {
+    generateJoinFieldDescriptorsFromViewAndJoins(layerModel, fields);
+  }
 
-    if (field.type === "String") {
-      field.length = 1024;
-    }
+  return fields;
+}
 
-    fields.push(field);
-  });
+function generateFieldDescriptorsFromDataSourcesArray(layerModel, serviceName) {
+  const fields = [];
 
-  if (layerModel.joins) {
-    layerModel.joins.forEach((dataSource) => {
+  var schema;
+  var view;
+  const primaryDataSource = layerModel.dataSources[0];
+  if (primaryDataSource.source === "view") {
+    schema = getSchema(layerModel.dataSources[0], serviceName);
+    view = layerModel.dataSources[0].view;
+    const viewDef = tde.getView(schema, view);
+    generateFieldDescriptorsFromViewDef(viewDef, fields);
+  } else if (primaryDataSource.source === "sparql") {
+    generateJoinFieldDescriptorsFromDataSource(primaryDataSource, fields);
+  }
+
+  if (layerModel.dataSources.length > 1) {
+    layerModel.dataSources.forEach((dataSource, index) => {
+      if (index < 1) return;  // skip first element since it is the primary source
       if (dataSource.fields) {
-        Object.keys(dataSource.fields).forEach((c) => {
-          const field = {
-            name : c,
-            type : getFieldType(dataSource.fields[c].scalarType)
-          };
-
-          if (field.type === "String") {
-            field.length = 1024;
-          }
-
-          fields.push(field);
-        });
+        generateJoinFieldDescriptorsFromDataSource(dataSource, fields);
+      } else {
+        const viewDef = tde.getView(dataSource.schema, dataSource.view);
+        generateFieldDescriptorsFromViewDef(viewDef, fields);
       }
     });
   }
 
   return fields;
 }
 
+function generateJoinFieldDescriptorsFromViewAndJoins(layerModel, fields) {
```

Comment:
Consider changing these functions to just return arrays of fields and then use the push() function with the ... spread operator to add the results to the array. It would look something like
```
const fields = [];
fields.push(...generateJoinFieldDescriptorsFromViewAndJoins(layerModel));
```

It's kind of a matter of preference but it makes the functions have no side effects so they are a bit cleaner.